app/HutchAgent/Services/CrateService.cs

```diff
@@ -0,0 +1,314 @@
+    {
+      Id = _publisherOptions.Name
+    });
+    crate.RootDataset.SetProperty("datePublished", DateTime.Now.ToString("yyyy-MM-dd'T'HH:mm:ssK"));
+
+    // Add dataset and files contained within and update the CreateAction
+    var createAction = crate.Entities.Values.First(x => x.GetProperty<string>("@type") == "CreateAction");
+    createAction.SetProperty("started", job.StartTime.ToString(CultureInfo.InvariantCulture));
+    createAction.SetProperty("ended", job.EndTime.ToString(CultureInfo.InvariantCulture));
+    createAction.SetProperty("actionStatus",
+      job.ExitCode == 0 ? "https://schema.org/CompletedActionStatus" : "https://schema.org/FailedActionStatus");
+    crate.Add(outputs);
+    foreach (var outputFile in outputFiles)
+    {
+      crate.Add(outputFile);
+      createAction.AppendTo("result", outputFile);
+    }
+
+    crate.Save(location: metaDirInfo.FullName);
+  }
+
+  /// <summary>
+  /// Update entity action status in an RO-Crate
+  /// </summary>
+  /// <param name="status">The status to update to</param>
+  /// <param name="entity">The RO-Crate entity to update</param>
+  public void UpdateCrateActionStatus(string status, Entity entity)
+  {
+    entity.SetProperty("actionStatus", new Part()
+    {
+      Id = status
+    });
+    _logger.LogInformation($"Set {entity.Id} actionStatus to {status}");
+  }
+
+  /// <summary>
+  /// Get execute entity from RO-Crate
+  /// </summary>
+  /// <param name="roCrate">the RO-Crate to look through</param>
+  /// <returns>The execute action entity</returns>
+  public Entity GetExecuteEntity(ROCrate roCrate)
+  {
+    //Get execution details
+    var mentions = roCrate.RootDataset.GetProperty<JsonArray>("mentions") ??
+                   throw new NullReferenceException("No mentions found in RO-Crate RootDataset Properties");
+    var mainEntity = roCrate.RootDataset.GetProperty<Part>("mainEntity") ?? throw new NullReferenceException();
+    // Check entity type and instrument
+    var executeEntityId = mentions.Where(mention => mention?["@id"] != null &&
+                                                    roCrate.Entities[
+                                                        mention["@id"]!.ToString() ??
+                                                        throw new NullReferenceException()]
+                                                      .Properties["@type"]!.ToString() == "CreateAction" &&
+                                                    roCrate.Entities[mention["@id"]!.ToString()]
+                                                      .Properties["instrument"]?["@id"]?.ToString() == mainEntity.Id)
+      .ToArray();
+    var executeAction =
+      roCrate.Entities[
+        executeEntityId.First()?["@id"]!.ToString() ??
+        throw new InvalidOperationException($"No entity found with id of {executeEntityId.First()?["@id"]}")];
+    _logger.LogInformation("Retrieved execution details from RO-Crate");
+    return executeAction;
```

Comment:
This is a bit messy. You could tidy it up by declaring the checks individually as variables above the `.Where` statement and then combine them in the `.Where` statement. That would look a bit cleaner and easier to read